influxdb3_write/src/write_buffer/flusher.rs

```diff
@@ -0,0 +1,245 @@
+        let (io_flush_notify_tx, io_flush_notify_rx) = std::sync::mpsc::channel();
+
+        let flusher = Self {
+            join_handle: Default::default(),
+            wal_io_handle: Default::default(),
+            shutdown_tx,
+            buffer_tx,
+        };
+
+        let wal_op_buffer_segment_state = Arc::clone(&segment_state);
+
+        *flusher.wal_io_handle.lock() = Some(
+            std::thread::Builder::new()
+                .name("write buffer io flusher".to_string())
+                .spawn(move || {
+                    run_io_flush(segment_state, io_flush_rx, io_flush_notify_tx);
+                })
+                .expect("failed to spawn write buffer io flusher thread"),
+        );
+
+        *flusher.join_handle.lock() = Some(tokio::task::spawn(async move {
+            run_wal_op_buffer(
+                wal_op_buffer_segment_state,
+                buffer_rx,
+                io_flush_tx,
+                io_flush_notify_rx,
+                shutdown_rx,
+            )
+            .await;
+        }));
+
+        flusher
+    }
+
+    pub async fn write_to_open_segment(
+        &self,
+        db_name: NamespaceName<'static>,
+        table_batches: HashMap<String, TableBatch>,
+        wal_op: WalOp,
+    ) -> crate::write_buffer::Result<WriteSummary> {
+        let (response_tx, response_rx) = oneshot::channel();
+
+        self.buffer_tx
+            .send(BufferedWrite {
+                wal_op,
+                database_write: DatabaseWrite::new(db_name, table_batches),
+                response_tx,
+            })
+            .await
+            .expect("wal op buffer thread is dead");
+
+        let summary = response_rx.await.expect("wal op buffer thread is dead");
+
+        match summary {
+            BufferedWriteResult::Success(summary) => Ok(summary),
+            BufferedWriteResult::Error(e) => Err(Error::BufferSegmentError(e)),
+        }
+    }
+}
+
+async fn run_wal_op_buffer(
+    segment_state: Arc<RwLock<SegmentState>>,
+    mut buffer_rx: mpsc::Receiver<BufferedWrite>,
+    io_flush_tx: std::sync::mpsc::Sender<Vec<WalOp>>,
+    io_flush_notify_rx: std::sync::mpsc::Receiver<wal::Result<SequenceNumber>>,
+    mut shutdown: watch::Receiver<()>,
+) {
+    let mut ops = Vec::new();
+    let mut write_batch = crate::write_buffer::buffer_segment::WriteBatch::default();
+    let mut notifies = Vec::new();
+    let mut interval = tokio::time::interval(BUFFER_FLUSH_INTERVAL);
+
+    loop {
+        // select on either buffering an op, ticking the flush interval, or shutting down
+        select! {
+            Some(buffered_write) = buffer_rx.recv() => {
+                ops.push(buffered_write.wal_op);
+                write_batch.add_db_write(buffered_write.database_write.db_name, buffered_write.database_write.table_batches);
+                notifies.push(buffered_write.response_tx);
+            },
+            _ = interval.tick() => {
```

Comment:
I think we want to skip missed ticks here so that we don't have the future complete multiple times to catch up and attempt to flush a lot.
https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip